### ISSUE TITLE
把模型变成推理服务：推理引擎

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#140]
# 模型部署成服务

## 推理引擎

将大模型部署成服务的库有 [ollama](https://ollama.com/)、[vLLM](https://docs.vllm.ai/en/stable/)、[TensorRT-LLM](https://nvidia.github.io/TensorRT-LLM/overview.html)、[TGI: Text-Generation-Interface](https://github.com/huggingface/text-generation-inference)，通过一些对比结论：vLLM比TGI性能优秀~15%，吞吐更高且服务更稳定。[^1] vLLM与TRT-LLM相比性能相当，但vLLM更节省资源。[^2]

### 如何兼容各类推理硬件（如NVIDIA GPU、国产GPU）

首先，是推理引擎来做兼容各类推理硬件的工作，其次，一般来说，不同硬件上的模型本身没有区别。

例如 vLLM 支持 ROCm、OpenVINO、CPU、Neuron、TPU、XPU，但在不同硬件上支持的能力略有不同。

### 使用 ollama

下载 ollama（v0.1.33已支持 llama3），在命令行执行 `ollama run llama3`，ollama会自动下载 llama3 的 8B 版本。

<img width="1208" alt="image" src="https://github.com/QingyaFan/blog/assets/10420579/d2af5a40-cabc-4ec3-a571-1b220687abdb">

<img width="1002" alt="image" src="https://github.com/QingyaFan/blog/assets/10420579/65409232-67b0-422a-b3c9-65f9b6cc9253">

可以看出来拉取模型时会下载几个文件，常见的有四个：

1. model：模型文件，包含了神经网络的结构和权重。模型是推理和生成任务的核心部分；
9. params：这些是模型的参数，通常包括网络中各层的权重和偏置。参数决定了模型在特定任务上的表现，它们是在训练过程中通过反向传播算法得到的；
10. template：模版一般用于定义模型的输入输出格式，或是提供一些示例和基础结构，帮助用户更好地使用模型；
11. license：模型使用的许可证信息，指明了在使用该模型时需要遵循的条款和条件。不同的许可证可能会对商业用途、修改和再分发等方面有不同的规定

### 使用 huggingface TGI

![image](https://github.com/user-attachments/assets/eae04a1b-b52e-4058-918b-9e85ef61be30)

Text Generation Interface（TGI）是部署和服务LLM的工具包，TGI 支持 Llama、Falcon等流行的开源LLM。TGI支持的特性：

- Simple launcher to serve most popular LLMs
- Production ready (distributed tracing with Open Telemetry, Prometheus metrics)
- Tensor Parallelism for faster inference on multiple GPUs
- Token streaming using Server-Sent Events (SSE)
- Continuous batching of incoming requests for increased total throughput
- Optimized transformers code for inference using [Flash Attention](https://github.com/HazyResearch/flash-attention) and [Paged Attention](https://github.com/vllm-project/vllm) on the most popular architectures
- Quantization with [bitsandbytes](https://github.com/TimDettmers/bitsandbytes) and [GPT-Q](https://arxiv.org/abs/2210.17323)
- [Safetensors](https://github.com/huggingface/safetensors) weight loading
- Watermarking with [A Watermark for Large Language Models](https://arxiv.org/abs/2301.10226)
- Logits warper (temperature scaling, top-p, top-k, repetition penalty)
- Stop sequences
- Log probabilities
- Fine-tuning Support: Utilize fine-tuned models for specific tasks to achieve higher accuracy and performance.
- [Guidance](https://huggingface.co/docs/conceptual/guidance): Enable function calling and tool-use by forcing the model to generate structured outputs based on your own predefined output schemas.

todo：都理解说的是什么

#### TGI 的可观测

对于大模型的可观测指标需要分为两类：软件、硬件
- 软件：例如推理耗时
- 硬件：例如GPU现存使用率

##### 软件

软件层面没有什么特殊，直接引入 open telemetry 导出要观察的指标即可。

```
pip install opentelemetry-api opentelemetry-sdk opentelemetry-instrumentation opentelemetry-exporter-otlp
```

```
from opentelemetry import trace
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from huggingface_hub import InferenceApi
from opentelemetry.exporter.otlp.proto.grpc.exporter import OTLPSpanExporter

# 配置Otel
resource = Resource.create({"service.name": "my-toy-inference-service"})
trace.set_tracer_provider(TracerProvider(resource=resource))

# 设置OTLP导出器
otlp_exporter = OTLPSpanExporter(endpoint="MY_OTLP_ENDPOINT", insecure=True)
trace.get_tracer_provider().add_span_processor(BatchSpanProcessor(otlp_exporter))

tracer = trace.get_tracer(__name__)

# 创建 HF 推理API
model = InferenceApi(repo_id="my-model-id")

def make_prediction(input_data):
    # start trace
    with tracer.start_as_current_span("model_inference"):
        # call model inference
        result = model(input_data)
        # end trace
        return result
    
# 示例输入数据
input_data = {
    "input": "Hello, how are you?",
}

# 调用推理函数
output = make_prediction(input_data)
print(output)
```

##### 硬件

todo [^3]

### llama3

llama 是生成式 AI 模型，[llama3](https://github.com/meta-llama/llama3) 是其最新版本（2024/4/29），相对于llama2，训练数据集是 15T token，llama2的 7 倍；支持 8K 的上下文长度，llama2的 2 倍。

目前 llama3 有 8B 和 70B 两个版本，也就是 80 亿参数和 700 亿参数，参数是权重参数，可以参考 #40 里的“什么是神经网络”节，也就是说神经元的量有 80 亿和 700 亿。

其性能也是有很大的提升，我们来解读一下 benchmark：

> https://huggingface.co/meta-llama/Meta-Llama-3-8B

<img width="642" alt="image" src="https://github.com/QingyaFan/blog/assets/10420579/106e4afb-1d33-43d7-8856-a60115aa9c9c">

- MMLU/GPQA/GSM-8K/MATH 是一些常见的模型 benchmark 测试数据集
- x-shot，表示每一类里有 x 个样本
- 得分越高越好

从测试结果看，llama3相对于llama2 有较大提升

## 客户端（如前端UI）

todo： use https://github.com/open-webui/open-webui

# 推理性能优化

- 模型量化（Model Quantization）：简单来说，模型量化就是通过某种方法将浮点模型转为定点模型。比如说原来的模型里面的权重（weight）都是float32，通过模型量化，将模型变成权重（weight）都是int8的定点模型。[^6]

- 模型剪枝（Model Pruning）：是一种用于减少神经网络模型参数数量和计算量的技术。它通过识别和去除在训练过程中对模型性能影响较小的参数或连接，从而实现模型的精简和加速。通常，模型剪枝可以分为两种类型：结构化剪枝（Structured Pruning）、非结构化剪枝（Unstructured Pruning）。[^7]

[^1]: https://tunehq.ai/blog/comparing-vllm-and-tgi
[^2]: https://pages.run.ai/hubfs/PDFs/Serving-Large-Language-Models-Run-ai-Benchmarking-Study.pdf
[^3]: https://hf-mirror.com/blog/martinigoyanes/llm-inference-at-scale-with-tgi
[^4]: https://www.apiseven.com/blog/what-is-sse
[^5]: https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events
[^6]: https://www.wehelpwin.com/article/4109
[^7]: https://cloud.baidu.com/qianfandev/topic/269647
[^8]: https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/features/optimizations/cpu_offloading.html
[^9]: https://docs.vllm.ai/en/latest/models/engine_args.html
[^10]: https://mp.weixin.qq.com/s/HXMLPM2KNkO6Mah-4btaEQ
